### ISSUE TITLE
lto link error undefined symbol: setjmp

### ISSUE THREAD

**Issue by calvin2021y** [WebAssembly/wasi-sdk#479]
I try build with `--target=wasm32-wasi -mllvm -wasm-enable-sjlj` and `-flto=full`

get this error:

```sh
undefined symbol: setjmp
undefined symbol: longjmp
nm test.o |grep jmp
         U longjmp
         U setjmp
```

If I remove `-flto=full`, all work as expect:

```sh
 nm test.o |grep jmp
         U __c_longjmp
         U __wasm_longjmp
         U saveSetjmp
         U testSetjmp
```

**Comment by yamt** [comment]
i guess you need `-Wl,-llvm,-wasm-enable-sjlj`.

**Comment by calvin2021y** [comment]
my link step not call from clang, so I add  `-mllvm -wasm-enable-sjlj` into wasm-ld.  this will work like `-Wl,-llvm,-wasm-enable-sjlj` ?


I get this error when link:

```sh
wasm-ld: error: /wasi/share/wasi-sysroot/lib/wasm32-wasi/llvm-lto/18.1.2-wasi-sdk/libsetjmp.a(rt.o): attempt to add bitcode file after LTO (__wasm_longjmp)
Error: wasm-ld failed with status: 1
```

**Comment by sbc100** [comment]
The options for a quick fix would be to either (1) remove `-flto` from you link command or (2) add `-Wl,-u__wasm_longjmp` (which would force that symbol to be included prior to LTO).

**Comment by calvin2021y** [comment]
after add `-Wl,-u__wasm_longjmp`,  the problem fixed.

any suggestion how to run the binary with `-wasm-enable-sjl`?  I try `wasm-interp` and wasmer, not work.

**Comment by yamt** [comment]
> my link step not call from clang, so I add `-mllvm -wasm-enable-sjlj` into wasm-ld. this will work like `-Wl,-llvm,-wasm-enable-sjlj` ?
> 
> I get this error when link:
> 
> ```shell
> wasm-ld: error: /wasi/share/wasi-sysroot/lib/wasm32-wasi/llvm-lto/18.1.2-wasi-sdk/libsetjmp.a(rt.o): attempt to add bitcode file after LTO (__wasm_longjmp)
> Error: wasm-ld failed with status: 1
> ```

this looks like a bug.
a fix: https://github.com/WebAssembly/wasi-libc/pull/529
for now, you can stop using lto as a workaround.

**Comment by yamt** [comment]
> after add `-Wl,-u__wasm_longjmp`, the problem fixed.
> 
> any suggestion how to run the binary with `-wasm-enable-sjl`? I try `wasm-interp` and wasmer, not work.

you can't run it.

* the setjmp runtime provided by wasi-sdk is not compatible with llvm 18.
   you need to use llvm 19 (an unreleased version) to build a working app.

* you need to use a runtime with exception-handling proposal. eg. toywasm, wamr classic interpreter.

**Comment by calvin2021y** [comment]
thanks for explain.

Is there a link I can download llvm19 sdk ?  (like nightly build results)

after `llvm 19` release,  can I run the wasm binary in browser ?

**Comment by yamt** [comment]
> thanks for explain.
> 
> Is there a link I can download llvm19 sdk ? (like nightly build results)
> 
> after `llvm 19` release, can I run the wasm binary in browser ?

if you are more interested in browsers, i guess it's better to try emscripten. (i'm not familiar with it.)

**Comment by sbc100** [comment]
Yes, emscripten has supported setjmp/longjmp for many years out-of-the-box.

**Comment by alexcrichton** [comment]
Since this issue was opened setjmp/longjmp is [documented how to use it](https://github.com/WebAssembly/wasi-sdk/blob/main/SetjmpLongjmp.md) and  with `-Wl,-mllvm,-wasm-enable-sjlj -Wl,-mllvm,-wasm-use-legacy-eh=false` locally this works for me, so I'm going to close this.
